### PR TITLE
Means

### PR DESCRIPTION
**Pull request by LukeMathWalker**
A basic implementation of arithmetic, harmonic and geometric mean.

Even though `ndarray` exposes `mean_axis` it does not provide `mean`, hence I have added it to the PR. Should I contribute it back to `ndarray` @jturner314?

### CODE REVIEW

File: src/summary_statistics/mod.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    ///
+    /// [`harmonic mean`]: https://en.wikipedia.org/wiki/Harmonic_mean
+    fn harmonic_mean(&self) -> Result<A, &'static str>
+        where
+            A: Float + FromPrimitive + PartialEq;
+
+    /// Returns the [`geometric mean`] `GM(X)` of all elements in the array:
+    ///
+    /// ```text
+    ///           n
```
Comment (jturner314) [inline src/summary_statistics/mod.rs]: I like big parentheses better in this case:

```rust
    ///         ⎛ n   ⎞¹⁄ₙ
    /// GM(X) = ⎜ ∏ xᵢ⎟
    ///         ⎝i=1  ⎠
```

File: src/summary_statistics/means.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+
+impl<A, S, D> SummaryStatisticsExt<A, S, D> for ArrayBase<S, D>
+where
+    S: Data<Elem = A>,
+    D: Dimension,
+{
+    fn mean(&self) -> Result<A, &'static str>
+    where
+        A: Clone + FromPrimitive + Add<Output=A> + Div<Output=A> + PartialEq + Zero
```
Comment (jturner314) [inline src/summary_statistics/means.rs]: What it the `PartialEq` bound used for?

File: src/summary_statistics/means.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                .expect("Converting number of elements to `A` must not fail.");
+            Ok(self.sum() / n_elements)
+        }
+    }
+
+    fn harmonic_mean(&self) -> Result<A, &'static str>
+    where
+        A: Float + FromPrimitive + PartialEq,
+    {
+        if self.len() == 0 {
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply src/summary_statistics/means.rs]: It was implemented separately to return a more precise error message, but if we use `Option` I can get rid of it.

File: src/summary_statistics/means.rs (hunk 2 above)
Comment (LukeMathWalker) [reply src/summary_statistics/means.rs]: Probably it survived some early iterations I was doing. I got rid of it.

**Comment by LukeMathWalker** [comment]
> I do think we should add .mean() to ndarray since ndarray already has .sum() and .mean_axis().

I agree - but the semantic of `mean_axis` is different at the moment: it panics instead of returning a `None` if the array is empty. That might have to be changed.

> It's unfortunate that harmonic_mean and geometric_mean allocate a temporary array for the result of the map. However, for the time being, I don't see a good way to avoid this while still taking advantage of the pairwise implementation of .sum() (rust-ndarray/ndarray#577).

It's exactly what I thought (and mean calculation is why I drafted the pairwise summation PR in the first place).

**Comment by LukeMathWalker** [comment]
CI on Rust 1.30 is failing with `error: Edition 2018 is unstable and only available for nightly builds of rustc.` for `noisy_floats`. Curious.

**Comment by jturner314** [comment]
Everything looks good, so I merged the PR. If `.mean()` gets added to `ndarray`, we can remove it from `ndarray-stats`.

> I agree - but the semantic of `mean_axis` is different at the moment: it panics instead of returning a `None` if the array is empty. That might have to be changed.

Will you please submit a PR to `ndarray` to add `.mean()`? I lean towards changing `mean_axis` to return `None` instead of making `mean` divide by zero, because of [this reasoning](https://github.com/rust-ndarray/ndarray/issues/512#issuecomment-433239806). However, for floating-point values, division by zero would be more convenient since it would just result in NaN instead of panicking. What do you think? I'll be interested to hear @bluss's opinion too.

**Comment by bluss** [comment]
@jturner314 Your rule makes sense, sounds good to me to return an Option. I think it's quite frequent anyway that you want to compute a mean and if empty, get zero, and the option makes it simple to do that.

**Pull request merged**